Release script: Added option to check for environment

spinscale opened the pull request:
In order to easily check for a correctly setup environment, you can
run

```
python3 dev-tools/build_release.py --check
```

It checks for env vars, needed commands and the correct java version.
Also supports colorful outputs.

A sample output looks like this

[![asciicast](https://asciinema.org/a/19326.png)](https://asciinema.org/a/19326?autoplay=1)

dadoonet commented:
w00t! Love your demo :)

Why we don't check MVN Java in check option?

spinscale commented:
@dadoonet added

dadoonet commented:
Just tested it locally and works really well.
LGTM.

dadoonet commented on dev-tools/build_release.py:
```diff
@@ -618,6 +618,47 @@ def check_norelease(path='src'):
+def check_environment_and_commandline_tools():
+  check_env_var('Checking for AWS env configuration AWS_SECRET_ACCESS_KEY_ID...     ', 'AWS_SECRET_ACCESS_KEY')
+  check_env_var('Checking for AWS env configuration AWS_ACCESS_KEY_ID...            ', 'AWS_ACCESS_KEY_ID')
+  check_env_var('Checking for SONATYPE env configuration SONATYPE_USERNAME...       ', 'SONATYPE_USERNAME')
```
This variable is not used in this script. Is this check done for another script? May be we could add a comment at the beginning of this script?

spinscale replied:
used for maven deploy

dadoonet commented on dev-tools/build_release.py:
```diff
@@ -618,6 +618,47 @@ def check_norelease(path='src'):
+  check_env_var('Checking for AWS env configuration AWS_SECRET_ACCESS_KEY_ID...     ', 'AWS_SECRET_ACCESS_KEY')
+  check_env_var('Checking for AWS env configuration AWS_ACCESS_KEY_ID...            ', 'AWS_ACCESS_KEY_ID')
+  check_env_var('Checking for SONATYPE env configuration SONATYPE_USERNAME...       ', 'SONATYPE_USERNAME')
+  check_env_var('Checking for SONATYPE env configuration SONATYPE_PASSWORD...       ', 'SONATYPE_PASSWORD')
```
Same here.

spinscale replied:
same here. :)

dadoonet commented on dev-tools/build_release.py:
```diff
@@ -618,6 +618,47 @@ def check_norelease(path='src'):
+  check_env_var('Checking for GPG env configuration GPG_KEY_ID...                   ', 'GPG_KEY_ID')
+  check_env_var('Checking for GPG env configuration GPG_PASSPHRASE...               ', 'GPG_PASSPHRASE')
+  check_env_var('Checking for S3 repo upload env configuration S3_BUCKET_SYNC_TO... ', 'S3_BUCKET_SYNC_TO')
+  check_env_var('Checking for git env configuration GIT_AUTHOR_NAME...              ', 'GIT_AUTHOR_NAME')
```
Same here

spinscale replied:
used by git when tags are created...

dadoonet commented on dev-tools/build_release.py:
```diff
@@ -618,6 +618,47 @@ def check_norelease(path='src'):
+  check_env_var('Checking for GPG env configuration GPG_PASSPHRASE...               ', 'GPG_PASSPHRASE')
+  check_env_var('Checking for S3 repo upload env configuration S3_BUCKET_SYNC_TO... ', 'S3_BUCKET_SYNC_TO')
+  check_env_var('Checking for git env configuration GIT_AUTHOR_NAME...              ', 'GIT_AUTHOR_NAME')
+  check_env_var('Checking for git env configuration GIT_AUTHOR_EMAIL...             ', 'GIT_AUTHOR_EMAIL')
```
Same here

dadoonet commented on dev-tools/build_release.py:
```diff
@@ -618,6 +618,47 @@ def check_norelease(path='src'):
+  # boto, check error code being returned
+  location = os.path.dirname(os.path.realpath(__file__))
+  command = 'python %s/upload-s3.py' % location
+  run_and_print('Testing boto python dependency...   ', partial(check_command_exists, 'python-boto', command))
```
I was not able to make this one becoming GREEN on my laptop though it works when I release plugins.

Basically, running manually:

```sh
python /path/to/elasticsearch/dev-tools/upload-s3.py
```

gives:

```
usage: upload-s3.py [-h] --file path to file [--bucket B42]
                    [--path elasticsearch/elasticsearch] [--key key]
upload-s3.py: error: argument --file/-f is required
```

Do you have a different output?

spinscale replied:
thats a bug and has been fixed now

spinscale commented:
@dadoonet fixed the python boto check, can you retry? worked on my linux system..

mikemccand commented:
Can we just always do `--check` by default so that when I try to release, it first checks and verifies I have the right env vars?

spinscale commented:
@mikemccand kept the `--check` parameter to exit early, but execute the checks all the time, also removed some redundant checks now

mikemccand commented:
Wonderful, thanks @spinscale!

rjernst commented:
Can the flag be called `--check-only`, to imply that is all it is doing? And this matches the var name used better anyways?

spinscale commented:
@rjernst good one, changed that as well

mikemccand commented on dev-tools/build_release.py:
```diff
@@ -72,6 +73,11 @@
+# console colors
+OKGREEN = '\033[92m'
+ENDC = '\033[0m'
+FAIL = '\033[91m'
```
Maybe rename to FAIL_COLOR, END_COLOR, OK_COLOR?

spinscale replied:
fixed

mikemccand commented on dev-tools/build_release.py:
```diff
@@ -618,6 +610,60 @@ def check_norelease(path='src'):
+    return False
+
+def check_environment_and_commandline_tools(check_only):
+  checks = list()
```
Why not []?

But also, why not just a boolean?  E.g. rename to `ok = True`, and then `ok = ok and check_env_var(..)` for each check?

mikemccand replied:
Hmm can't do `ok and ...` because Python will short-circuit and won't print the failures...

Maybe just:

```
if check_env_var(...):
  ok = False
```

for each check?

Or leave it as a list :)

spinscale replied:
left it as list for now :)

mikemccand commented on dev-tools/build_release.py:
```diff
@@ -636,9 +682,12 @@ def check_norelease(path='src'):
 
   parser.set_defaults(dryrun=True)
   parser.set_defaults(smoke=None)
+  parser.set_defaults(check=False)
```
Shouldn't this be `check_only` now?

spinscale replied:
true. fixed.

mikemccand commented:
LGTM, thanks @spinscale!

spinscale committed:
Release script: Always check for valid environment

In order to be sure that a release can be executed on the local machine,
the build_release script now checks for environment variables and tries
to execute a  couple of commands.

In order to easily check for a correctly setup environment, you can
run the following commands, which exits early and does not trigger a
release process.

```
python3 dev-tools/build_release.py --check-only
```
